cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/ProgramStorePublisher.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.store;
+
+import co.cask.cdap.app.runtime.Arguments;
+import co.cask.cdap.app.runtime.ProgramStateWriter;
+import co.cask.cdap.app.store.RuntimeStore;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.service.Retries;
+import co.cask.cdap.common.service.RetryStrategies;
+import co.cask.cdap.proto.BasicThrowable;
+import co.cask.cdap.proto.ProgramRunStatus;
+import co.cask.cdap.proto.id.ProgramId;
+import com.google.common.base.Supplier;
+import org.apache.twill.api.RunId;
+
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * An implementation of the ProgramStateWriter that persists directly to the store
+ */
+public final class ProgramStorePublisher implements ProgramStateWriter {
```

Comment:
Usually an implementation has the interface name as suffix. E.g. `DirectStoreProgramStateWriter`.